read-fonts/src/tables/gsub/closure.rs

```diff
@@ -241,247 +368,475 @@ impl GlyphClosure for LigatureSubstFormat1<'_> {
+        match self {
+            ContextFormat2::Plain(table) => table.coverage(),
+            ContextFormat2::Chain(table) => table.coverage(),
+        }
+    }
+
+    fn input_class_def(&self) -> Result<ClassDef<'_>, ReadError> {
+        match self {
+            ContextFormat2::Plain(table_ref) => table_ref.class_def(),
+            ContextFormat2::Chain(table_ref) => table_ref.input_class_def(),
+        }
+    }
+
+    fn rule_sets(&self) -> impl Iterator<Item = Option<Result<Format2RuleSet, ReadError>>> {
+        let (left, right) = match self {
+            ContextFormat2::Plain(table) => (
+                Some(
+                    table
+                        .class_seq_rule_sets()
+                        .iter()
+                        .map(|rs| rs.map(|rs| rs.map(Format2RuleSet::Plain))),
+                ),
+                None,
+            ),
+            ContextFormat2::Chain(table) => (
+                None,
+                Some(
+                    table
+                        .chained_class_seq_rule_sets()
+                        .iter()
+                        .map(|rs| rs.map(|rs| rs.map(Format2RuleSet::Chain))),
+                ),
+            ),
+        };
+        left.into_iter()
+            .flatten()
+            .chain(right.into_iter().flatten())
+    }
+}
+
+impl Format2RuleSet<'_> {
+    fn rules(&self) -> impl Iterator<Item = Result<Format2Rule, ReadError>> {
+        let (left, right) = match self {
+            Format2RuleSet::Plain(table) => (
+                Some(
+                    table
+                        .class_seq_rules()
+                        .iter()
+                        .map(|rule| rule.map(Format2Rule::Plain)),
+                ),
+                None,
+            ),
+            Format2RuleSet::Chain(table) => (
+                None,
+                Some(
+                    table
+                        .chained_class_seq_rules()
+                        .iter()
+                        .map(|rule| rule.map(Format2Rule::Chain)),
+                ),
+            ),
+        };
+        left.into_iter()
+            .flatten()
+            .chain(right.into_iter().flatten())
+    }
+}
+
+//https://github.com/fonttools/fonttools/blob/a6f59a4f87a0111/Lib/fontTools/subset/__init__.py#L1215
+impl GlyphClosure for ContextFormat2<'_> {
+    fn add_reachable_glyphs(&self, ctx: &mut ClosureCtx) -> Result<(), ReadError> {
+        let coverage = self.coverage()?;
+        let Some(cur_glyphs) = intersect_coverage(&coverage, ctx.current_glyphs()) else {
+            return Ok(());
+        };
+
+        let classdef = self.input_class_def()?;
+        let our_classes = make_class_set(ctx.glyphs(), &classdef);
+        for (class_i, seq) in self
+            .rule_sets()
             .enumerate()
-            .filter_map(|(i, seq)| seq.filter(|_| input_classes.contains(i as u16)))
+            .filter_map(|(i, seq)| seq.map(|seq| (i as u16, seq)))
```

Comment:
should we filter ruleSets by coverage first? see: https://github.com/fonttools/fonttools/blob/a6abbac37b2ef272904a9d8134272901138ea8a9/Lib/fontTools/subset/__init__.py#L1217